### PR TITLE
feat (sequel): Initial Sequel instrumentation

### PR DESCRIPTION
**Pull request by jrgns**
OpenTelemetry Instrumentation for the [Sequel Database Toolkit](https://sequel.jeremyevans.net/).

This can be used instead of DB specific instrumentations.

### CODE REVIEW

**Comment by arielvalentin** [comment]
@jrgns Thank you for your contribution! 

I will try to give this a review later this week but I have a few questions for you before we begin the process:

Are you a member of the OTel GitHub Organization? 

Are you willing to commit to providing on-going maintenance for this instrumentation? E.g. Review PRs, Fix Vulnerabilities, Maintain Version Compatibility, etc...

**Comment by jrgns** [comment]
Hey @arielvalentin 

No, I'm not part of the OTel Github Org

Yes. I'm willing to commit to maintenance etc.

**Comment by arielvalentin** [comment]
@jrgns looks like the generator added the unreleased instrumentation gem as a dependency in all. 

Please remove it from all an I'll create a follow up PR to include it once we have a release of this gem published in rubygems.

**Comment by jrgns** [comment]
@arielvalentin check it now?

**Comment by arielvalentin** [comment]
@jrgns I will take a look tomorrow during US business hours

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: Why did you call this `component`? Is this supposed to be used as a tag to find all spans related to this instrumentation?

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:26]: 🚨 Recording the Raw SQL statement will leak PII and other kinds of sensitive data. 

If the `sequel` gem does not support a way to provide a templated SQL statement then we must omit it. 

@kaylareopelle has been working on a PR to extract reusable SQL obfuscation code here: https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/529

Once that is merged, then you may update this instrumentation to optionally use it.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:31]: Does the Sequel gem wrap lower level driver libraries or should the naming here follow Database Semantic Conventions?

https://opentelemetry.io/docs/specs/semconv/database/database-spans/

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:45]: I this best done here or as part of the `instrumentation.install` method?
```suggestion
  Database.register_extension(:opentelemetry, self)
```

File: instrumentation/sequel/opentelemetry-instrumentation-sequel.gemspec
Comment (arielvalentin) [inline instrumentation/sequel/opentelemetry-instrumentation-sequel.gemspec:38]: If this is the minimum dependency then this should enforce the minimal version check using a `compatible` check. 

E.g. https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/instrumentation.rb#L23

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:27]: This must correspond to the type of SQL statement being performed, i.e. `SELECT`, `INSERT`, etc...

https://opentelemetry.io/docs/specs/semconv/database/database-spans/#call-level-attributes

**Review by arielvalentin — CHANGES_REQUESTED** [review]
I have left a few comments for you to review with some requested changes. Please be patient as I am not familiar with the Sequel gem and will need your help digging into some specifics about how extensions work. 

Thanks again for this contribution!

File: instrumentation/sequel/README.md
Comment (jrgns) [reply instrumentation/sequel/README.md:48]: I tried using `Sequel.extension(:opentelemetry)`, but that doesn't seem to register it to all the DB instances. So you have to run it on the result of a `Sequel.connect`.

I'll check if there's a better way to do it, but for now this is the workaround

File: instrumentation/sequel/opentelemetry-instrumentation-sequel.gemspec
Comment (jrgns) [reply instrumentation/sequel/opentelemetry-instrumentation-sequel.gemspec:38]: It is the minimum version. Will add the version check.

File: instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb
Comment (jrgns) [reply instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb:20]: Yes, will add the check.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:15]: Ok. Do we change the file name to `open_telemetry.rb` as well? I know gems like zeitwerk won't like the mismatch.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: Yes. Is there a better semantic naming for that?

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:26]: I suspect the template is available. Will find it.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:15]: Unless I am mistaken, zeitwerk should not be impacted by this since it should not be hot/reloading external gems loaded by bundler.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:31]: It's a wrapper.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb]: No template available. Removed the sql.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb]: Removed this as well since it's not available for all Sequel adapters

**Comment by jrgns** [comment]
@arielvalentin I updated what I could. I had to remove some of the info as discussed.

(btw, do I resolve the conversations, or do you re-review and then resolve?)

**Review by kaylareopelle — COMMENTED** [review]
Hi @jrgns! 👋 It's great to see sequel instrumentation coming to OpenTelemetry! Thanks for your work! 

I left a few comments in the code. I've been working on some new instrumentation lately and added references to my new gem in a few additional places to help with CI and releases. You may want to do this too. 

Check out:
* .github/workflows/ci-instrumentation-canary.yml
* .github/workflows/ci-instrumentation.yml
* .toys/.data/releases.yml
* instrumentation/all/lib/opentelemetry/instrumentation/all.rb
* instrumentation/all/opentelemetry-instrumentation-all.gemspec

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: I do not know off the top of my head. Would you mind reaching out to the folks in CNCF or opening an issue https://github.com/open-telemetry/semantic-conventions/issues? 

https://cloud-native.slack.com/archives/C01QZFGMLQ7
https://cloud-native.slack.com/archives/C041APFBYQP

**Comment by jeremyevans** [comment]
I don't have any open-telemetry experience.  However, from a brief review of the extension, you probably should wrap `execute_insert` in addition to `execute_dui`.

**Comment by arielvalentin** [comment]
Thank you @jeremyevans for taking the time out to help with this review! 🙇🏼

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (arielvalentin) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: On second thought, i think that this should really already be covered by `InstrumentationScope`. 

That is a resource attribute attached to all spans and is currently named `OpenTelemetry::Instrumentation::Sequel`. 

Would that be sufficient?

Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: image
Yep, that's fine.

File: instrumentation/sequel/README.md
Comment (jrgns) [reply instrumentation/sequel/README.md:48]: @jeremyevans since you've gotten involved. What's the best way to automatically register this extension on all current and future DB connections?

File: instrumentation/all/Gemfile
Comment (jrgns) [reply instrumentation/all/Gemfile:14]: To prevent it from being included when adding all the instrumentation until this has been completed.

**Comment by jrgns** [comment]
> I don't have any open-telemetry experience. However, from a brief review of the extension, you probably should wrap `execute_insert` in addition to `execute_dui`.

Done

File: instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb
Comment (jrgns) [reply instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb:18]: I tried `Sequel::Database.extension(:extension_name)` but it doesn't have the desired effect. Will check your other suggestion.

**Comment by jrgns** [comment]
@kaylareopelle thanx for the suggestions 😄  I'm working through them now.

I've taken a look at your gems. You'll see that I removed the logging of SQL from this gem (because of lack of templating in Sequel). I'm wondering if that's not the best way forward - leave the templating / obfuscation to the library gems (like pg and mysql) and keep the Sequel instrumentation to a minimum?

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb]: Or we can look at @kaylareopelle 's SQL Obfuscation gem?

File: instrumentation/sequel/README.md
Comment (jrgns) [reply instrumentation/sequel/README.md]: Updated.

File: instrumentation/sequel/README.md
Comment (jrgns) [reply instrumentation/sequel/README.md]: Updated.

File: instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb
Comment (jrgns) [reply instrumentation/sequel/lib/opentelemetry/instrumentation/sequel/instrumentation.rb:18]: The `each` executed on each DB connection works nicely. Thanx!

**Comment by jrgns** [comment]
@arielvalentin I've addressed all the comments, including great suggestions from @kaylareopelle . Anything else, or can we get this merged?

**Comment by arielvalentin** [comment]
@jeremyevans Although I know you are not familiar with OTel Tracing, your expertise and review of the optimal usage of the Sequel extensions in this PR would be greatly appreciated 🙇🏼. 

> @arielvalentin I've addressed all the comments, including great suggestions from @kaylareopelle . Anything else, or can we get this merged?

@jrgns Yes! Something that would be helpful would be some screenshots of the instrumentation in action. 

I am also working through the failing tests that are unrelated to your PR that have come up after the Rails 7.1 release. I would like to get those fixes merged prior to merging this PR.

I appreciate your patience as we work through the blockers before being able to give this another look.

**Comment by jeremyevans** [comment]
> @jeremyevans Although I know you are not familiar with OTel Tracing, your expertise and review of the optimal usage of the Sequel extensions in this PR would be greatly appreciated 🙇🏼.

I briefly reviewed and provided comments earlier.  There is little Sequel-specific code in the PR, the vast amount of it is OpenTelemetry boilerplate that wouldn't benefit from my review.

My only additional comment would be that the tests should include update, delete, and insert queries.

**Comment by jrgns** [comment]
@arielvalentin something like this?

image

image

**Comment by jrgns** [comment]
@arielvalentin @jeremyevans added tests for update, delete and insert queries (and fixed the original tests).

File: instrumentation/sequel/README.md
Comment (jeremyevans) [reply instrumentation/sequel/README.md]: `Sequel::Database.extension` will add it to future Database objects created, and you can go through `Sequel::DATABASES` to add it to existing Database objects.  However, that should not be done by a library.  You should ask the user to explicitly enable this per Database object.  If a user uses multiple Database objects in their application, they may not want telemetry added to all of them.  At the very least, adding it automatically to existing and future Database objects should be opt-in.

Comment (jrgns) [reply instrumentation/sequel/README.md]: Thanx @jeremyevans , that clarifies it.

@arielvalentin happy that we turn off the automatic enabling of telemetry, and allow the user to either

- manually add telemetry to a connection or further connections by running `Sequel::Database.extension(:opentelemetry)` or `DB.extension(:opentelemetry)`
- specify the `target` option when configuring Opentelemetry to be one of `current`, `future` or `all` (where `all` will add telemetry to current and future connections)
?

**Comment by kaylareopelle** [comment]
@jrgns  - Sorry about the delay here! Re-reviewing your PR now.

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (kaylareopelle) [inline instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb:15]:
```suggestion
```

**Comment by kaylareopelle** [comment]
> @kaylareopelle thanx for the suggestions 😄 I'm working through them now.
> 
> I've taken a look at your gems. You'll see that I removed the logging of SQL from this gem (because of lack of templating in Sequel). I'm wondering if that's not the best way forward - leave the templating / obfuscation to the library gems (like pg and mysql) and keep the Sequel instrumentation to a minimum?

@jrgns - I like the idea of leaving the templating/obfuscation to the library gems. Let's keep the code there and not worry about it in our Sequel instrumentation. 😄

**Review by kaylareopelle — CHANGES_REQUESTED** [review]
I found one stray bit of commented-out code. Other than that, this looks good to me!

**Comment by kaylareopelle** [comment]
@jrgns - My apologies, I forgot to follow up on the CI/release documents in my earlier review.

Could you add your gem to the following files?
```
.github/workflows/ci-instrumentation-canary.yml
.github/workflows/ci-instrumentation.yml
.toys/.data/releases.yml
```

This should allow your `sequel` suite to run as part of the CI checks and will allow the gem to eventually be released. Unfortunately, these file updates aren't included automatically in the instrumentation generator.

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (arielvalentin) [inline instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb:29]: The tests do not seem very specific to me or help me understand as an end user what I should expect to see on a span when I use this. 

The tests for `traces dataset executes` e.g. may continue to pass even though the span does not have the data we expect. 

May I ask that you expand on these tests (although it may seem verbose) to include assertions around the spans that were captured?

E.g. name, kind, attributes, if any events where captured etc..

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (jrgns) [reply instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb]: Sorted

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:17]: I didn't realise it's memoized there. Will remove the memoization.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (jrgns) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:28]: It seems to match the names of other spans?

image

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (jrgns) [reply instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb]: I've made some changes, but will expand it even further.

**Comment by jrgns** [comment]
@kaylareopelle I've added the gem to the CI files.

**Comment by jrgns** [comment]
@arielvalentin anything else you need on this?

File: instrumentation/sequel/CHANGELOG.md
Comment (arielvalentin) [inline instrumentation/sequel/CHANGELOG.md:5]: The release automation generates the changelog entries for you. It requires this file be empty for the initial release. 

```suggestion
```

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (arielvalentin) [inline instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb:32]: Please use predicate assertions.

These will be more explicit and print the contents of the `finished_spans` in the error message as opposed to the cryptic `expected 0 but was 1`

```suggestion
      _(exporter.finished_spans).must_be_empty
```

File: instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb
Comment (arielvalentin) [inline instrumentation/sequel/test/opentelemetry/instrumentation/sequel_test.rb:39]: Similar to the test above, you will be able to test both the size of the array and the contents at once by comparing objects instead of using size properties and navigating the objects. 

```suggestion
      _(exporter.finished_spans.map(&:name)).must_equal(['sequel.execute'])
```

I would recommend applying this change to similar tests in this file.

**Review by arielvalentin — CHANGES_REQUESTED** [review]

**Review by kaylareopelle — CHANGES_REQUESTED** [review]

**Comment by jrgns** [comment]
@kaylareopelle @arielvalentin  changes made.

**Comment by kaylareopelle** [comment]
Thanks, @jrgns! It looks like [the sequel tests are failing with JRuby](https://github.com/open-telemetry/opentelemetry-ruby-contrib/actions/runs/6684240506/job/18412661724?pr=672). The CI skips JRuby for Active Record, so perhaps it should be skipped for Sequel too. 

On [Sequel's homepage](https://sequel.jeremyevans.net/), it seems like the gem should be compatible with JRuby, but I think we might to install the JBDC adapter test environment to make it work.

@arielvalentin, I'm not aware of the OTel Ruby policy that drives JRuby testing for instrumentation. Can you provide some guidance here?

**Comment by arielvalentin** [comment]
It's mostly a capacity limitation. 

I think we really want to support the JRuby test but none of the existing maintainers use it, so trying to maintain the integration test suite is quite difficult.  

If you are up to the challenge of enabling JRuby integration tests then I would welcome your contribution! https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/4

**Comment by simi** [comment]
tests can be started when sqlite3 is removed from gemspec and added to gemfile

```
gem 'sqlite3', '~> 1.6.4', platform: :ruby
gem 'jdbc-sqlite3', '< 3.42', platform: :jruby
```

But tests are still failing due to some difference. I'm looking into.

File: instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb
Comment (simi) [inline instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: `opts[:database]` is `nil` for jdbc connection like `jdbc:sqlite::memory:`. @jeremyevans is there any better way in here to get database name? I have checked `https://sequel.jeremyevans.net/rdoc/classes/Sequel/Database.html` and I wasn't able to find out.

Comment (jeremyevans) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: Sequel treats JDBC connection string URIs as opaque, it does not try to parse them.  You could try parsing the JDBC URI yourself if you want to find the database name.  Note that it isn't unusual to not have a database name even in the non-JDBC case.  `postgres:///?user=foo` will use the default database for user `foo` on PostgreSQL, but `opts[:database]` will still be nil in that case.

Comment (jeremyevans) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: I think the most robust way to handle this is to use `Database#database_type` to determine the type of database in use, then run database-specific code to get the database name using a query.  However, Sequel does not have built-in code to do that, and trying to handle it for all databases that Sequel supports seems challenging.

Considering that approach isn't really feasible, parsing the JDBC URI and returning the path part as the database name is probably the best you can do.  At least that gets you behavior similar to other adapters.

Comment (simi) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: I did some experiments, but it is not simple to extract the DB name from JDBC URI in unified way and also it is not possible to redact password in JDBC URI. Considering all this together, would it be acceptable to make this one optional @arielvalentin @kaylareopelle and call `compact` on attributes?

PS: It would be possible to add `DB_JDBC_DRIVER_CLASSNAME`, I can open additional PRO to enhance.

Comment (arielvalentin) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: It's a bit slower than checking if values are nil before adding them to the hash but using `compact!` is fine by me.

Comment (arielvalentin) [reply instrumentation/sequel/lib/sequel/extensions/opentelemetry.rb:23]: Yes that's fine. Instrumentations should do their best to include attributes.

**Comment by jrgns** [comment]
What's the next steps here?

**Comment by arielvalentin** [comment]
1. Please update your branch and address any merge conflicts.
2. Fix any failing tests/builds
3. Re-request a review from folks whose comments you've addressed

**Comment by solutus** [comment]
Hello, is there any plans to maintain Sequel now?

**Comment by arielvalentin** [comment]
> Hello, is there any plans to maintain Sequel now? 

Not at the moment. 

We don't have anyone who is willing to be a long term maintainer of the instrumentation. 

Is that something you'd be interested in helping with?

**Comment by jrgns** [comment]
I lost track of this, but happy to take it on again if there's interest?

**Comment by arielvalentin** [comment]
None of our current maintainers or approvers are Sequel gem users but we are happy to support you if you're able to help us maintain it. 

Please let us know how you'd like to proceed.

**Pull request closed without merging**
